### PR TITLE
Draft HLD - Image management

### PR DESCRIPTION
**Pull request by arunsundark**

### CODE REVIEW

File: doc/mgmt/SONiC-Management-Framework-Image-mgmt-HLD.md [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+# 2 Functionality
+## 2.1 Target Deployment Use Cases
+
+**image install**
+
+This command is used to install a new image on the alternate image partition. This command takes a path to an installable SONiC image or URL and installs the image.
+
+**show image-list**
+
+This command displays information about currently installed images. It displays a list of installed images, currently running image and image set to be loaded in next reboot.
```
Comment (mikelazar) [inline doc/mgmt/SONiC-Management-Framework-Image-mgmt-HLD.md:76]: How many installed images can there be in total ? Would it be possible to mention this here ?
Based on the description on the previous command ("image install") there seem to be only two... But here we talk about a list...

Comment (arunsundark) [reply doc/mgmt/SONiC-Management-Framework-Image-mgmt-HLD.md:76]: Sonic does not enforce a hard limit on the number of images that can be installed. I think space can be the only limiting factor here.  But if you want give a guideline value for the number of images that can be installed, I can add that here as well.

File: doc/mgmt/SONiC-Management-Framework-Image-mgmt-HLD.md [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+
+# Revision
+| Rev  |    Date    |      Author       | Change Description |
+| :--: | :--------: | :---------------: | ------------------ |
+| 0.1  | 09/13/2019 | Arunsundar Kannan | Initial version    |
+
+# About this Manual
+This document provides general information about the Image management and installation feature implementation in SONiC.
+# Scope
+Covers Northbound interface for the Image management feature, as well as Unit Test cases.
```
Comment (jeffyin-dell) [inline doc/mgmt/SONiC-Management-Framework-Image-mgmt-HLD.md]: Need to be more consistent about capitalization here. 
My suggestion:
Covers northbound interface for the Image Management feature, as well as unit test cases.

File: doc/mgmt/SONiC-Management-Framework-Image-mgmt-HLD.md [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+| ---------------- | ------------------------------ |
+| Image Management | Image installation, validation |
+
+# 1 Feature Overview
+
+Provide management framework capabilities to handle:
+
+- Image Installation
+- List available images
+- Removal of an image
```
Comment (jeffyin-dell) [inline doc/mgmt/SONiC-Management-Framework-Image-mgmt-HLD.md]: Suggest "Removal of an available image"

File: doc/mgmt/SONiC-Management-Framework-Image-mgmt-HLD.md [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+- Set default image for next boot
+
+
+## 1.1 Requirements
+
+### 1.1.1 Functional Requirements
+
+Provide management framework support to existing SONiC capabilities with respect to Image management
```
Comment (jeffyin-dell) [inline doc/mgmt/SONiC-Management-Framework-Image-mgmt-HLD.md]: Please keep capitalization consistent for Image Management.

File: doc/mgmt/SONiC-Management-Framework-Image-mgmt-HLD.md [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+
+### 1.1.2 Configuration and Management Requirements
+- CLI style configuration and show commands
```
Comment (jeffyin-dell) [inline doc/mgmt/SONiC-Management-Framework-Image-mgmt-HLD.md]: Remove "style"

File: doc/mgmt/SONiC-Management-Framework-Image-mgmt-HLD.md [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+- REST API support
+- gNMI Support
+
+### 1.1.3 Scalability Requirements
+N/A
+### 1.1.4 Warm Boot Requirements
+
```
Comment (jeffyin-dell) [inline doc/mgmt/SONiC-Management-Framework-Image-mgmt-HLD.md]: N/A?

File: doc/mgmt/SONiC-Management-Framework-Image-mgmt-HLD.md [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+## 1.2 Design Overview
+### 1.2.1 Basic Approach
+The feature will use the traditional translib/transformer approach with an addition of a service running in the host which does all the "actual work". The service will use the existing sonic installer script in the host.
```
Comment (jeffyin-dell) [inline doc/mgmt/SONiC-Management-Framework-Image-mgmt-HLD.md]: Define "actual work".
Describe more specifically what is meant by the "traditional translib/transformer approach." I think you'll also need to elaborate on RPC support and how that is supported by translib, or at least how transformer will handle translations from OpenConfig gNOI to operations on the container/host.

File: doc/mgmt/SONiC-Management-Framework-Image-mgmt-HLD.md [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+
+**image remove**
+
+This command is used to remove the unused SONiC image from the disk. Note that it's *not* allowed to remove currently running image.
+
+Note: The above text is taken from the sonic command line reference Wiki (https://github.com/Azure/sonic-utilities/blob/master/doc/Command-Reference.md#sonic-installer)
+
+## 2.2 Functional Description
+
+Upon receiving a request, the app module forwards the request to the service running at the host, which calls the backend implementation to do all image management related tasks.
```
Comment (jeffyin-dell) [inline doc/mgmt/SONiC-Management-Framework-Image-mgmt-HLD.md]: You may want to provide a simple diagram for this.

File: doc/mgmt/SONiC-Management-Framework-Image-mgmt-HLD.md [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+
+## 3.4 SyncD
+N/A
+
+## 3.5 SAI
+N/A
+
+## 3.6 User Interface
+### 3.6.1 Data Models
+TBD (Working on a custom Sonic yang).
```
Comment (jeffyin-dell) [inline doc/mgmt/SONiC-Management-Framework-Image-mgmt-HLD.md]: What about https://github.com/openconfig/gnoi/blob/master/os/os.proto ?

File: doc/mgmt/SONiC-Management-Framework-Image-mgmt-HLD.md [hunk 10]
```diff
@@ -0,0 +1,225 @@
...
+
+### 3.6.2 CLI
+#### 3.6.2.1 Configuration Commands
+
+**image install**
```
Comment (jeffyin-dell) [inline doc/mgmt/SONiC-Management-Framework-Image-mgmt-HLD.md:170]: Please provide a plain-English description for each command

File: doc/mgmt/SONiC-Management-Framework-Image-mgmt-HLD.md [hunk 11]
```diff
@@ -0,0 +1,225 @@
...
+
+# 7 Warm Boot Support
+
+TBD
+
+# 8 Scalability
+N/A
+
+# 9 Unit Test
+List unit test cases added for this feature including warm boot.
```
Comment (jeffyin-dell) [inline doc/mgmt/SONiC-Management-Framework-Image-mgmt-HLD.md:241]: Needs a table of unit test cases

File: doc/mgmt/SONiC-Management-Framework-Image-mgmt-HLD.md [hunk 12]
```diff
@@ -0,0 +1,225 @@
...
+
+# 10 Internal Design Information
+Internal BRCM information to be removed before sharing with the community.
```
Comment (jeffyin-dell) [inline doc/mgmt/SONiC-Management-Framework-Image-mgmt-HLD.md]: Delete this line

**Review by jeffyin-dell — CHANGES_REQUESTED** [review]

File: doc/mgmt/SONiC-Management-Framework-Image-mgmt-HLD.md (hunk 9 above)
Comment (arunsundark) [reply doc/mgmt/SONiC-Management-Framework-Image-mgmt-HLD.md]: We are not going to use that model. We will be using a custom yang with RPC. gNMI will use the same yang, not the protobuf.

File: doc/mgmt/SONiC-Management-Framework-Image-mgmt-HLD.md (hunk 1 above)
Comment (babukr) [reply doc/mgmt/SONiC-Management-Framework-Image-mgmt-HLD.md:76]: As per https://github.com/Azure/SONiC/wiki/SONiC-to-SONiC-update, SONiC supports two images at a time. When you install a newer image when there are two images, the oldest one is replaced.

File: doc/mgmt/SONiC-Management-Framework-Image-mgmt-HLD.md [hunk 13]
```diff
@@ -0,0 +1,250 @@
...
+- Set default image for next boot
+
+
+## 1.1 Requirements
+
+### 1.1.1 Functional Requirements
+
+Provide management framework support to existing SONiC capabilities with respect to Image Management
+
+### 1.1.2 Configuration and Management Requirements
```
Comment (kkvedula) [inline doc/mgmt/SONiC-Management-Framework-Image-mgmt-HLD.md:49]: Provide a list of REST API based on the data models supported.

File: doc/mgmt/SONiC-Management-Framework-Image-mgmt-HLD.md [hunk 14]
```diff
@@ -0,0 +1,250 @@
...
+N/A
+
+## 3.4 SyncD
+N/A
+
+## 3.5 SAI
+N/A
+
+## 3.6 User Interface
+### 3.6.1 Data Models
```
Comment (kkvedula) [inline doc/mgmt/SONiC-Management-Framework-Image-mgmt-HLD.md:131]: Is this based on SONIC YANG?

File: doc/mgmt/SONiC-Management-Framework-Image-mgmt-HLD.md [hunk 15]
```diff
@@ -0,0 +1,250 @@
...
+    |  +--ro output
+    |     +--ro status?          int32
+    |     +--ro status-detail?   string
+    +---x image-list
+       +--ro output
+          +--ro images-found?   string
+```
+
+### 3.6.2 CLI
+#### 3.6.2.1 Configuration Commands
```
Comment (kkvedula) [inline doc/mgmt/SONiC-Management-Framework-Image-mgmt-HLD.md:168]: Can a command to boot to ONIE prompt on next reboot be added?

**Review by jeffyin-dell — APPROVED** [review]

File: doc/mgmt/SONiC-Management-Framework-Image-mgmt-HLD.md (hunk 13 above)
Comment (arunsundark) [reply doc/mgmt/SONiC-Management-Framework-Image-mgmt-HLD.md:49]: Done

File: doc/mgmt/SONiC-Management-Framework-Image-mgmt-HLD.md (hunk 11 above)
Comment (jeffyin-dell) [reply doc/mgmt/SONiC-Management-Framework-Image-mgmt-HLD.md:241]: Approval pending list of unit test cases

File: doc/mgmt/SONiC-Management-Framework-Image-mgmt-HLD.md (hunk 14 above)
Comment (arunsundark) [reply doc/mgmt/SONiC-Management-Framework-Image-mgmt-HLD.md:131]: Yes

File: doc/mgmt/SONiC-Management-Framework-Image-mgmt-HLD.md (hunk 15 above)
Comment (arunsundark) [reply doc/mgmt/SONiC-Management-Framework-Image-mgmt-HLD.md:168]: Sure, we will try to add it but after buzznik

**Pull request merged**